### PR TITLE
Check win conditions

### PR DESCRIPTION
**Pull request by LiveNathan**
Initially I solved it with nested for each loops to check surround locations by column and row with an additional while loop inside. That was working so I asked ai to refactor and it suggested an array of tuples, which I like because it allows me to remove some of the location checking conditions and I think it should be faster by checking in one direction at a time.

### CODE REVIEW

File: main.js [hunk 1]
```diff
@@ -11,13 +11,18 @@ const BOARD_SIZE = 14;
...
-const SPHERE_HEIGHT_SEGMENT = SPHERE_WIDTH_SEGMENT / 2;
+const SPHERE_HEIGHT_SEGMENT = SPHERE_WIDTH_SEGMENT / 3;
+const WINNING_NUMBER_OF_STONES = 5;
 
 const scene = new THREE.Scene();
 let currentPlayer = 'black';
+let currentPlayerLabel = null;
+let winnerAnnouncement = null;
+let restartButtonLabel = null;
+let stonesInScene = [];
```
Comment (iansedano) [inline main.js]: Nothing wrong with this global state, especially at this level of complexity, but often what you'll see is some namespacing to keep certain pieces of global state together, i.e.

```suggestion
const GAME_STATE = {
  let currentPlayer : 'black',
  let currentPlayerLabel : null,
  let winnerAnnouncement: null,
  let restartButtonLabel : null,
  let stonesInScene : []
}
```

This comes in handy later if you want to implement "undo" or "history" type functionality, as you can save all the different stages of the game state in an array, for example.

File: main.js [hunk 2]
```diff
@@ -11,13 +11,18 @@ const BOARD_SIZE = 14;
...
 
 function initializeGameBoard() {
-    const gameBoard = Array(BOARD_SIZE).fill().map(() => Array(BOARD_SIZE).fill(null));
+    const gameBoard = Array(BOARD_SIZE + 1).fill().map(() => Array(BOARD_SIZE + 1).fill(null));
```
Comment (iansedano) [inline main.js:27]: I'm seeing the same changes in both PRs, which isn't ideal for review. I forgot about some of the dyanmics involved when using multiple branches in this way, so it's my fault. Let's chat about it at the next meeting

File: main.js [hunk 3]
```diff
@@ -210,6 +274,79 @@ function animate(renderer, labelRenderer, camera, controls) {
     labelRenderer.render(scene, camera);
 }
 
+function clearGameBoard(scene, gameBoard) {
```
Comment (iansedano) [inline main.js:275]: Nicely scoped function. Looks like its a very appropriate abstraction around a well defined task, with a clear name. These are all nice indicators of the "right" abstraction, if such a thing exists!

File: main.js [hunk 4]
```diff
@@ -210,6 +274,79 @@ function animate(renderer, labelRenderer, camera, controls) {
...
+        clearGameBoard(scene, gameBoard);
+        currentPlayer = 'black';
+        currentPlayerLabel.element.textContent = `Player turn: ${currentPlayer}`;
+
+        if (winnerAnnouncement) {
+            scene.remove(winnerAnnouncement);
+            winnerAnnouncement = null;
+        }
+    }
+}
```
Comment (iansedano) [inline main.js]: It looks like this is a factory function. That is, it creates an event handler. If that's the intention, it's typical to include that in the name. Maybe `restartGameHandlerFactory`

**Review by iansedano — APPROVED** [review]
Awesome! It works! Great job in coming up with a couple of approaches to solve the "check the winner" problem. I tried a bunch of edge cases (diagonals, edges, mainly) , but couldn't break it, so great work!

I'm still having to click various times to get the stone to be placed, are you not getting that same behavior?

**Comment by LiveNathan** [comment]
> Awesome! It works! Great job in coming up with a couple of approaches to solve the "check the winner" problem. I tried a bunch of edge cases (diagonals, edges, mainly) , but couldn't break it, so great work!
> 
> I'm still having to click various times to get the stone to be placed, are you not getting that same behavior?

Unsure. I haven't run into the same issue. I click, wait for the stone to appear, then click again and it works.

**Pull request merged**